### ISSUE TITLE
Enable dsub from Terra notebooks

### ISSUE THREAD

**Issue by mbookman** [DataBiosphere/saturn-ui#998]
Today users are not able to run dsub from Terra notebooks.
This issue was first filed directly against Leonardo here:

- https://github.com/DataBiosphere/leonardo/issues/542

We need to take the next steps in Terra.

If documentation is correct for [pipelines.run](https://cloud.google.com/genomics/reference/rest/v2alpha1/pipelines/run#authorization-scopes) then we need to add either of:

- https://www.googleapis.com/auth/cloud-platform
- https://www.googleapis.com/auth/genomics

to notebook VMs launched through Terra.

This assumes also that:

- The Genomics API is enabled within the Terra workspace Cloud project
- The user's pet service account has requisite underlying grants to call the Pipelines API

The document grants to call [pipelines.run](https://cloud.google.com/genomics/reference/rest/v2alpha1/pipelines/run#authorization-scopes) and [operations.list](https://cloud.google.com/genomics/reference/rest/v2alpha1/projects.operations/list) are:

- genomics.operations.create
- genomics.operations.list

Note that in the original issue filed against Leonardo, deflaux@ indicated wanting to run both dsub as well as tensorflow. I don't know the scopes or permissions needed for tensorflow. We could file that as a separate issue.

**Comment by bradtaylor** [comment]
As Matt says, let's proceed by adding the cloud-plaftorm scope and testing with dsub to see if this is sufficient.

Marking this issue as high-risk so as to trigger a security review

**Comment by kleighwe** [comment]
See details on the Leo side here: https://github.com/DataBiosphere/leonardo/issues/542

**Comment by kleighwe** [comment]
@bradtaylor  can provide a "hello world" dsub

**Comment by rtitle** [comment]
Hello, 

I'd like to provide an update on this. On the Leo side we now provide the ability to set explicit scopes on clusters. I recently created a cluster with the `https://www.googleapis.com/auth/cloud-platform` scope and tried to test whether dsub works.

**First I tried to use the `local` provider:**
```
%%bash
/home/jupyter-user/.local/bin/dsub \
   --provider local \
   --logging /tmp/dsub-test/logging/ \
   --output OUT=/tmp/dsub-test/output/out.txt \
   --command 'echo "Hello World" > "${OUT}"' \
   --wait
```
This failed -- Nicole indicates that it's due to it creating a docker within a docker, which doesn't work in this context.

**Then I tried to use the `google` provider which uses Pipelines API V1:**
```
%%bash
# TODO Terra workspace bucket name should be available via an environment variable.
/home/jupyter-user/.local/bin/dsub \
  --project $GOOGLE_PROJECT \
  --zones 'us-central1-*' \
  --provider google \
  --logging gs://rtitle_test/dsub/logging/ \
  --output OUT=gs://rtitle_test/dsub/output/out.txt \
  --command 'echo "Hello World" > "${OUT}"' \
  --wait
```
This failed with the error message:
```
"Error in job echo--jupyter-user--190227-164843-56 - code 7: Required 'compute.zones.list' permission for 'projects/broad-firecloud-dsde'"
```
**Then I tried to use the `google-v2` provider which uses Pipelines API V2:**
```
%%bash
# TODO Terra workspace bucket name should be available via an environment variable.
/home/jupyter-user/.local/bin/dsub \
  --project $GOOGLE_PROJECT \
  --zones 'us-central1-*' \
  --provider google \
  --logging gs://rtitle_test/dsub/logging/ \
  --output OUT=gs://rtitle_test/dsub/output/out.txt \
  --command 'echo "Hello World" > "${OUT}"' \
  --wait
```
This failed with:
```
HttpError 403 when requesting https://genomics.googleapis.com/v2alpha1/pipelines:run?alt=json returned "The caller does not have permission"
```

So none of them work currently. :) I think we want to decide what functionality we want to support.
1. Local -- This might be beneficial to users to run jobs on their clusters managed through Leo/Terra. There is probably some docker shenanigans we can do to allow this. This work would fall on the Leo team.
2. PAPI v1 -- We should probably not worry about this as it's end-of-life and Terra prod is soon moving to PAPI v2 anyway.
3. PAPI v2 -- As mentioned we are releasing support for PAPI v2 in Q1 (this work is being done by Cloud Accounts). When this is deployed we should be granting the `v2alpha1` permission. So it might Just Work once PAPI v2 support is released to prod. However if there are additional gaps between the GCP roles that dsub requires and what Terra provides then we would need to do some Rawls work to grant the appropriate roles when adding users to the can-compute policy. This doc is helpful in explaining policies and roles: https://docs.google.com/document/d/1uOGtmw_l_Ve2gqJohLGntJmQz69md2h3T2EAvyKTm20/edit#

**Comment by mbookman** [comment]
Thanks @rtitle !

The goal would be to support the `google-v2` provider.
The `google` provider is deprecated and the `local` provider would not work in this context, as @deflaux  indicated, due to docker-in-docker.

With supporting the `google-v2` provider, we would like to look at two different use-cases:

1- I have my own Cloud project and want to run workflows there.
2- I want to run workflows in the Terra project.

If we could enable setting the `cloud-platform` scope on the cluster, then @deflaux and I could test (1) to help inform (2).

**Comment by bradtaylor** [comment]
@rtitle it sounds like cloud-platform scope is a necessary prerequisite whether they are trying to run in a Terra project or their own, so I think we should do this now anyway. Do you concur?

In parallel we should look into what must be done to ensure dsub works when PAPI V2 support lands. If possible, we should ask to add additional requirements to that work so we get dsub as part of our efforts.

**Comment by rtitle** [comment]
Yes on both counts @bradtaylor.

Regarding the current status of PAPI v2, the dev work is ready and we are waiting on government approval (for Google Cloud KMS) before releasing to prod. But in the meantime we could validate dsub functionality in a test environment. The best way to do this would probably be to spin up a fiab with the following branches and test notebook functionality from there:
cromwell: dev
rawls: mtalbott-private-auto-test
sam: mtalbott-grant-kms-permissions
fc-develop: mtalbott-papiv2

**Comment by mbookman** [comment]
Should I create a separate issue specifically for tracking the ability to set the `cloud-platform` scope when I create a cluster? As it stands today, in addition to `dsub`, I also cannot run [gcp-variant-transforms](https://github.com/googlegenomics/gcp-variant-transforms) among other tools. When I do, I get:

```
ERROR: (gcloud.alpha.genomics.pipelines.run) PERMISSION_DENIED: Request had insufficient authentication scopes.
```

Is there something specifically standing in the way of having Terra UI allow users to specify this?

Thanks!

**Comment by bradtaylor** [comment]
@mbookman I agree. Since cloud-platform scope is itself insufficient for dsub, but also useful for other things, we should throw this back in the backlog and resurrect #1035. I will do that now

**Comment by bradtaylor** [comment]
When #1035 is merged, this issue should only be blocked on PAPI-V2

**Comment by mbookman** [comment]
The opening up of the cloud-platform scope should allow one to run dsub in a Terra notebook, so long as the dsub call specifies a non-Terra Cloud project:

```
dsub \
  --provider google-v2 \
  --project my-non-terra-project \
  <etc>
```

The Terra access control model, which is at the workspace level, does not mesh well with the Pipelines API access control model, which is at the project level. The only way to have these sync up today would be to have the restriction that every workspace is in its own project.

I'm working on a design for a new `dsub` provider `terra`, which would send submissions to the Terra (firecloud) [submissions API](https://api.firecloud.org/#/Submissions). If this works out, this approach would allow user to send requests to the submissions API using `dsub` and then in addition to using `dstat` and `ddel`, users could go monitor their submissions just like existing Terra workflows.
